Use Vector for EventHandler

eatkins opened the pull request:
ArrayList::add is not thread safe. I ran into cases where async tests
using utests would fail even when all of the individual tests passed.
This was because multiple threads called back into the handle method of
the handler instance variable, which just delegated to eventList::add.
When this happened, one of the events would get added to the list as a
null reference, which would manifest as an NPE upstream on the master
process. After this change, my tests stopped failing.

(See the guidelines for contributing, linked above)

eatkins committed:
Use ConcurrentLinkedDeque for EventHandler

ArrayList::add is not thread safe. I ran into cases where async tests
using utests would fail even when all of the individual tests passed.
This was because multiple threads called back into the handle method of
the handler instance variable, which just delegated to eventList::add.
When this happened, one of the events would get added to the list as a
null reference, which would manifest as an NPE upstream on the master
process. After this change, my tests stopped failing.

dwijnand reviewed (approved):
lgtm! nice fix, @eatkins!

two thoughts:

1. I think we could merge this into 1.1.x too for 1.1.2
2. would it be possible to construct a test for this? (e.g with async tests in utest)

dwijnand commented:
/cc @retronym who's also been dealing with concurrency issues in sbt's testing.

retronym commented:
Yep, https://github.com/sbt/sbt/pull/3997 addresses a similar problem in non-forked tests.

retronym commented:
But I also recently discovered #4010 in the context of Scalacheck forked tests, which has the same symptom you're reporting: a single test failure doesn't fail the entire suite. That wasn't a race condition, but a problem of losing data by clobberring a binding in a map of results, when the test framework created many nested `Task`s for fine grained parallel execution (each test methods in a single class got a parallel `Task`, but because these shared the fully-qualified-name of the enclosing class the result aggregation code only ended up using the status of the final test)

@eatkins How sure are you that this change (alone) fixes the problem you saw?

retronym commented on testing/agent/src/main/java/sbt/ForkMain.java:
```diff
@@ -294,7 +295,7 @@ private void runTestTasks(final ExecutorService executor, final Task[] tasks, fi
 					Task[] nestedTasks;
 					final TaskDef taskDef = task.taskDef();
 					try {
-						final List<ForkEvent> eventList = new ArrayList<ForkEvent>();
```
Alternatively:

```
final Collection<ForkEvent> eventList = new java.util.concurrent.ConcurrentLinkedDeque<ForkEvent>();
```

eatkins replied:
I like this if for no other reason than that it makes it very explicit that we expect that the eventList may be touched by multiple threads.

retronym commented on testing/agent/src/main/java/sbt/ForkMain.java:
```diff
@@ -294,7 +295,7 @@ private void runTestTasks(final ExecutorService executor, final Task[] tasks, fi
 					Task[] nestedTasks;
 					final TaskDef taskDef = task.taskDef();
 					try {
-						final List<ForkEvent> eventList = new ArrayList<ForkEvent>();
```
I agree this is racy.

eatkins committed:
Add test for async utest TestSuites

Sometimes when utest runs async tests (i.e. tests that return a future)
the test suite will fail even when none of the individual tests do. This
is due to a data race in sbt. Most, but not all, of the time, this test
will induce that race.

eatkins commented:
I added a test to the PR that induces the data race that I was seeing.

@retronym I think my issue was slightly different than #4010. In my case, all of my individual tests passed, but the overall suite was failing.

eed3si9n reviewed (changes requested):
LGTM
I'd also vote to rebase this to 1.1.x.

eed3si9n commented:
It seems https://github.com/sbt/sbt/pull/4030 does just that.

dwijnand commented:
closing in favour of #4030
